src/v/kafka/server/connection_context.cc

```diff
@@ -64,24 +64,77 @@ class invalid_virtual_connection_id : public std::exception {
     explicit invalid_virtual_connection_id(ss::sstring msg)
       : _msg(std::move(msg)) {}
 
+    const char* what() const noexcept final { return _msg.c_str(); }
+
 private:
     ss::sstring _msg;
 };
 
-bytes parse_virtual_connection_id(const kafka::request_header& header) {
+// Tuple containing virtual connection id and client id. It is returned after
+// parsing parts of virtual connection id.
+struct virtual_connection_client_id {
+    virtual_connection_id v_connection_id;
+    std::optional<std::string_view> client_id;
+};
+
+vcluster_connection_id parse_vcluster_connection_id(std::string_view str) {
+    vcluster_connection_id cid;
+    std::stringstream sstream(str.data());
+    sstream >> std::hex >> cid;
+    return cid;
+}
```

Comment:
I think there should be some validation to ensure that the vcluster_connection_id contains only valid hex characters.  The streaming operation won't throw any exception if an invalid character is encountered and will simply stop processing.